underdocs-version/src/main/kotlin/underdocs/version/retriever/DefaultVersionRetriever.kt

```diff
@@ -0,0 +1,20 @@
+package underdocs.version.retriever
+
+import underdocs.version.Version
+import underdocs.version.reader.VersionPropertiesReader
+import java.util.*
+
+class DefaultVersionRetriever: VersionRetriever {
+    override fun get(): Version {
+        val versionProperties : Properties by lazy {
+            val propertiesReader = VersionPropertiesReader()
```

Comment:
Instead of explicitly creating a `VersionPropertiesReader` instance, the class should have a property of type `PropertiesReader` which is injected via the constructor.

~~~~Kotlin
class DefaultVersionReader(private val propertiesReader: PropertiesReader): VersionRetriever {
}
~~~~

The current approach has two problems:

  * It defeats the purpose of creating interfaces and abstractions. By directly using the `VersionPropertiesReader` class, this class will directly depend on an implementation instead of depending on an interface.
  * By explicitly calling the constructor of `VersionPropertiesReader`, one cannot unit test this class in isolation. This is the result of a dependency not being injected, but created.

Reply:
Ok, will try to keep these in mind and maybe look into unit testing as well.

Reply:
As we will go into the plugin refactoring, you'll have plenty of opportunities to practice unit testing!